lib/std/mem.zig

```diff
@@ -1593,21 +1593,57 @@ test containsAtLeast {
 /// T specifies the return type, which must be large enough to store
 /// the result.
 pub fn readVarInt(comptime ReturnType: type, bytes: []const u8, endian: Endian) ReturnType {
-    var result: ReturnType = 0;
+    const bits = @typeInfo(ReturnType).Int.bits;
+    const signedness = @typeInfo(ReturnType).Int.signedness;
+    const WorkType = std.meta.Int(signedness, @max(16, bits));
+    var result: WorkType = 0;
     switch (endian) {
         .big => {
             for (bytes) |b| {
                 result = (result << 8) | b;
             }
         },
         .little => {
-            const ShiftType = math.Log2Int(ReturnType);
+            const ShiftType = math.Log2Int(WorkType);
             for (bytes, 0..) |b, index| {
-                result = result | (@as(ReturnType, b) << @as(ShiftType, @intCast(index * 8)));
+                if (index * 8 >= bits) break;
+                result = result | (@as(WorkType, b) << @as(ShiftType, @intCast(index * 8)));
             }
         },
     }
-    return result;
+    return @as(ReturnType, @truncate(result));
+}
+
+test readVarInt {
+    try testing.expect(readVarInt(u0, &[_]u8{}, .big) == 0x0);
+    try testing.expect(readVarInt(u0, &[_]u8{}, .little) == 0x0);
+    try testing.expect(readVarInt(u8, &[_]u8{0x12}, .big) == 0x12);
+    try testing.expect(readVarInt(u8, &[_]u8{0xde}, .little) == 0xde);
+    try testing.expect(readVarInt(u16, &[_]u8{ 0x12, 0x34 }, .big) == 0x1234);
+    try testing.expect(readVarInt(u16, &[_]u8{ 0x12, 0x34 }, .little) == 0x3412);
+
+    try testing.expect(readVarInt(i8, &[_]u8{0xff}, .big) == -1);
+    try testing.expect(readVarInt(i8, &[_]u8{0xfe}, .little) == -2);
+    try testing.expect(readVarInt(i16, &[_]u8{ 0xff, 0xfd }, .big) == -3);
+    try testing.expect(readVarInt(i16, &[_]u8{ 0xfc, 0xff }, .little) == -4);
+
+    // Return type can be undersized (bytes.len * 8 > @typeInfo(ReturnType).Int.bits)
```

Comment:
this contradicts the documented API of the function:

```
/// T specifies the return type, which must be large enough to store
/// the result.
```

Does this PR fix a bug or change the intended behavior of the API? If the latter, we should also change the API docs in this PR.

@squeek502 i see you had thoughts on this in #20440.

Reply:
> Does this PR fix a bug or change the intended behavior of the API?

Undersized return types were allowed before this PR as well:

```
const std = @import("std");

pub fn main() void {
    const slice = &[_]u8{ 0xFF, 0xFF, 0xFF, 0xFF };
    const val = std.mem.readVarInt(u25, slice, .little);
    std.debug.print("{}\n", .{val});
}
```

outputs `0x1FFFFFF` (silently truncating 7 bits)

but it doesn't work for *all* undersized types, e.g. using `u24` or smaller in the test panics at runtime with `shift amount is greater than the type size` or `integer cast truncated bits`.

This PR is focused only on making `readVarInt` compile when using int types <= 8 bits wide.

Reply:
So then what behavior are you proposing for this function? (and for `readInt`?) What change to this API are you proposing:

```
/// Reads an integer from memory with size equal to bytes.len.
/// T specifies the return type, which must be large enough to store
/// the result.
```

Reply:
> What change to this API are you proposing:

I'm not sure yet. Either one of:

- Runtime assertion that `@typeInfo(ReturnType).Int.bits >= bytes.len` (this would technically be a breaking change)
- Continue to allow undersized types, but make the function always invoke checked illegal behavior if the *value* will not fit in the type (i.e. enforce there are no truncated bits)

Hence why I think the options should be considered in a followup issue rather than this PR.

`readInt` is unaffected, since it enforces at compile-time that the type's bits are divisible by 8 and the number of bytes matches the type exactly (since it takes a pointer to an appropriately sized array).

Reply:
> (but note that it doesn't work for all undersized types, e.g. using u24 or smaller in this example panics at runtime with shift amount is greater than the type size or integer cast truncated bits)

I'll note here that this PR now also supports this (no more panic). If you all want this to be removed, then I'll do another commit.

Reply:
i'll defer to someone with more of a stake in this issue to review the code. i chimed in to participate, but don't wait for me to approve it. my interest is mostly in specifying the intended behavior, whatever that behavior may be.

Reply:
Made an issue for the undersized types problem: https://github.com/ziglang/zig/issues/20521